Add cilium-health endpoints for datapath connectivity probing

joestringer opened the pull request:
This series adds support for creating an extra endpoint by default which runs `cilium-health`. The goal of this is to allow operators to be able to query the local cilium-health daemon and understand whether the end-to-end path from the local node to endpoints on a remote node are working as expected. In this series, the endpoint is spun up with no policy applied, so it should simply validate that when operating through the endpoint BPF forwarding logic, the endpoints are accessible. Subsequent PRs are expected to tackle the multiple paths that may be taken through a cilium-controlled node, such as via Services or through some form of policy, particular L7 Policies when a userspace proxy is involved.

This series is split roughly into
* Various cleanups necessary to create an endpoint from the daemon
* Net namespace spawn script
* Instantiation of the cilium-health endpoint and attachment of this to Cilium
* Extending `pkg/node` and `pkg/k8s` to forward cilium-health IPs to other nodes
* Extending the cilium `status` API to export cilium-health IPs
* Extending cilium-health to perform probing to other nodes.

Next steps:
* [x] Test the k8s multi-node code, make sure that cilium-health instances ping other nodes
* [x] Most runtime tests fail because they don't expect the extra endpoint in the list
* [x] Resolve "endpoint restore" for cilium-health endpoint
* [x] Add ginkgo test for cilium-health
* [x] Restrict API calls to manipulate reserved labels on endpoints

Related: #1947

joestringer committed:
daemon: Spawn and run cilium-health as an endpoint

Add support for running an additional, passive instance of cilium-health
inside a network namespace, interacting as though it is an endpoint. The
new endpoint will have the labels "reserved:health".

Subsequent commits will propagate this new endpoint's IP into pkg/node
so that the cilium-health instance colocated with cilium may acquire
these endpoint IPs and perform periodic probing to these endpoints.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Shift socket removal to launcher

An upcoming commit will open a passive cilium-health daemon on HTTP
only, so don't remove the unix socket by default in that daemon; only if
it is launched directly by Cilium.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Fix typo in error messages

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Add spawn_netns.sh script

This script allows a network namespace to be created and a command run
within that namespace. The caller must specify the name, veth pair, ip
addresses, command and arguments to be run in the namespace. The script
will create the namespace, configure the devices including IP addresses
within the namespace, and run the command in a manner that will restart
the target command when interrupted or killed. Killing the script itself
will interrupt the command and clean up the namespace, restoring it all
to the state prior to script invocation.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Refactor endpoint creation.

This chunk of code in the daemon can be moved into endpointmanager for
better reuse by upcoming commits.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
pkg: Add simple pidfile library

Add a simple pidfile library for reuse in upcoming commits, and to
remove the vendor dependency.

Requested-by: Alexander Alemayhu <alexander@alemayhu.com>
Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
common: Add new SetupVethWithNames() helper

This will allow the upcoming cilium-health net namespace to get devices
with more intuitive names.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
common: Support transforming route to ip command

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
endpoint: Add UpdateSecLabels to Owner

This will allow some code to be factored outside of the daemon, but
still retain the core of the ability for UpdateSecLabels to
differentiate different error conditions for generating API responses.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
k8s: Add health IP annotations for node

Extend the k8s annotations to include two new annotations, for storing
the IPv4 and IPv6 addresses of the cilium-health endpoint in the node.
All nodes may then fetch all other nodes' health endpoint IPs for
probing purposes.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
node: Set cilium-health endpoint IPs in node

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
api: Add health endpoints to node API

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
api: Regenerate autogenerated API code

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Probe health endpoint IPs

This commit begins to probe health endpoint IPs and report their
results.

Signed-off-by: Joe Stringer <joe@covalent.io>

aalemayhu commented on cilium-health/cmd/root.go:
```diff
@@ -131,7 +135,17 @@ func initConfig() {
+			os.Exit(-1)
+		}
+	}
 
```
Could you move this pidfile creation into a separate package? Would be nice to get rid of the `facebookgo/pidfile ` dependency I added to the agent.

joestringer replied:
Sure.

joestringer replied:
Done.

joestringer committed:
cilium-health: Add --pidfile option

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,216 @@
+// LaunchAsEndpoint launches the cilium-health agent in a nested network
+// namespace and attaches it to Cilium the same way as any other endpoint,
+// but with special reserved labels.
+func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressing, opts *option.BoolOptions) context.CancelFunc {
```
In practice it's invoked like `health.LaunchAsEndpoint()`, so the above doesn't apply.

aanm replied:
I don't think he understands you

raybejjani commented on bpf/spawn_netns.sh:
```diff
@@ -0,0 +1,108 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+NETNS=$1
```
Could you add a short description of how/when/what this script does? For extra credit, also document expectations for each parameter

aanm replied:
+1

raybejjani commented on daemon/endpoint.go:
```diff
@@ -691,7 +655,7 @@ func (d *Daemon) UpdateSecLabels(id string, add, del labels.Labels) middleware.R
 	}
 
-	return nil
+	return 0, nil
```
There is a `PutEndpointIDLabelsOKCode` that might be more appropriate, since we treat this int return as a HTTP status code otherwise.

raybejjani commented on daemon/endpoint.go:
```diff
@@ -600,20 +565,20 @@ func (h *getEndpointIDHealthz) Handle(params GetEndpointIDHealthzParams) middlew
-func (d *Daemon) UpdateSecLabels(id string, add, del labels.Labels) middleware.Responder {
+func (d *Daemon) UpdateSecLabels(id string, add, del labels.Labels) (int, error) {
 	addLabels, _ := labels.FilterLabels(add)
 	delLabels, _ := labels.FilterLabels(del)
```
I realise it was already like this, but should we handle the errors here? now that we return an error

joestringer replied:
Previously we just handed back an API error and the caller would send that out. Now, there's another path, being called from within the daemon. In this case it could be handled either inside here or in the caller. However, for the case that this command is triggered from an API call, we still need to forward the error back on out.

I've left this as-is for the moment.

raybejjani commented on daemon/endpoint.go:
```diff
@@ -141,74 +140,40 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
 	if len(add) > 0 {
-		errLabelsAdd := h.d.UpdateSecLabels(params.ID, add, labels.Labels{})
+		err, errLabelsAdd := h.d.UpdateSecLabels(params.ID, add, labels.Labels{})
 		if errLabelsAdd != nil {
```
Should the variables here be `HTTPCode, err := h.d.UpdateSecLabels(params.ID, add, labels.Labels{})`
I spent a while trying to figure out why err wasn't being handled, but the message was! Then I noticed err was an int which was even odder. We make similar calls in at least one other spot, so if you choose to change this one we'd have to change the others.

raybejjani commented on daemon/endpoint.go:
```diff
@@ -600,20 +565,20 @@ func (h *getEndpointIDHealthz) Handle(params GetEndpointIDHealthzParams) middlew
 	}
 	if ep == nil {
-		return NewPutEndpointIDLabelsNotFound()
+		return PutEndpointIDLabelsNotFoundCode, fmt.Errorf("")
```
I imagine this is nil since not found isn't an OK case, and something somewhere might check on the error (so it can't be nil). If so, we should just have `"Endpoint ID not found"` as the error message in case it is printed by a caller, someday.

raybejjani replied:
Oh, we do exactly this on line 600.

raybejjani commented on daemon/state_test.go:
```diff
@@ -202,3 +202,7 @@ func (ds *DaemonSuite) TestSyncLabels(c *C) {
 }
+
+func (ds *DaemonSuite) UpdateSecLabels(id string, add, del labels.Labels) (int, error) {
+	return 500, fmt.Errorf("Not implemented")
```
To my surprise, there is a 501 Not Implemented code! https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501

raybejjani commented on pkg/policy/identity.go:
```diff
@@ -191,17 +191,22 @@ const (
 	ReservedIdentityCluster
+
+	// ReservedIdentityHealth represents the local cilium-health endpoint
+	ReservedIdentityHealth
```
Should we also add this to labels.go and make it possible to reference them in policies? I'm not sure how that code interacts with tracing and our other tooling (like monitor) but I can imagine someone explicitly wanting to block health. I might be crazy, though, I'm not sure.

joestringer replied:
These endpoints are intended to be reserved for cilium debug use, and when writing this tool I've been thinking about how to answer questions like:

* Does basic node-to-node connectivity work? (already merged)
* Does basic container-to-container connectivity work? (this commit)
* Does connectivity work when there's an allow policy in place? (future)
* Does connectivity work when packets are sent through a proxy? (future)

For the last two items above, I expect that we'll want to reference these endpoints in policies. However, I'd expect these to be managed by cilium and not by end users.

The defaults of this tool are set up with the idea in mind that you don't have to turn something on, or deploy something, etc. when you're hitting a connectivity issue. The daemons are running, you just need to query them and interpret the output.

With regards to users wanting to explicitly blocking health, I think that a configuration option on the daemon would be the best option to address this, so we don't spawn these extra daemons on startup. This shouldn't be too much work, but at this stage I plan to defer it until it is requested.

raybejjani replied:
> For the last two items above, I expect that we'll want to reference these endpoints in policies. However, I'd expect these to be managed by cilium and not by end users.

Yeah, I was mostly thinking along the lines of not having special cases in the internal policy resolution. We'd deny user policies from having one thing or the other during validation. I'm not sure how much work this would create, though.

 > With regards to users wanting to explicitly blocking health, I think that a configuration option on the daemon would be the best option to address this, so we don't spawn these extra daemons on startup. This shouldn't be too much work, but at this stage I plan to defer it until it is requested.

How would that manifest in a health probe? if nothing is listening the connection would be refused. Would the report distinguish in some way? Since this option would be an agent level thing, but it may not correlate with what pods are scheduled on that node it might be a little confusing? I haven't thought through it.

joestringer replied:
>> With regards to users wanting to explicitly blocking health, I think that a configuration option on the daemon would be the best option to address this, so we don't spawn these extra daemons on startup. This shouldn't be too much work, but at this stage I plan to defer it until it is requested.

> How would that manifest in a health probe? if nothing is listening the connection would be refused. Would the report distinguish in some way? Since this option would be an agent level thing, but it may not correlate with what pods are scheduled on that node it might be a little confusing? I haven't thought through it.

I'd expect that remote instances that still have health enabled would observe that the "health-disabled" node exists and try to ping it, but the ping would fail. If you wanted to distinguish deliberate configuration vs. unexpected connectivity loss, the "health-disabled" instance would have to propagate the fact that its health is deliberately disabled to the other cilium instances, expose it to cilium-health, which could then print the output differently for those nodes.

raybejjani commented on daemon/main.go:
```diff
@@ -657,6 +635,10 @@ func runDaemon() {
 
+	addressing := d.getNodeAddressing()
+	cancelHealth := health.LaunchAsEndpoint(d, addressing, d.conf.Opts)
+	defer cancelHealth()
```
It took me a long time to understand the distinction between `d.ciliumHealth.Run()` and `health.LaunchAsEndpoint()`, mostly because I didn't understand what cilium-health was doing. I could understand what each call did, but not the intent behind them. I'm not sure how to make it more obvious overall, but perhaps comments here explaining what each is and that the `.Run()` spawned one tests endpoint connectivity against the `.LaunchAsEndpoint` spawned one might help a bit. Alternatively, more comments on the intended usecases for each function in it's comment might also work, although that seems like the wrong place to put that.

joestringer replied:
As a brief description here: We launch two instances of cilium-health, one on the host and one which acts as an endpoint. This allows us to test multiple connectivity paths to other nodes in the cluster.

With respect to the code, I can add brief comments here to distinguish the two and I'll think about an appropriate place to put a description like the above.

aanm replied:
Yes please

raybejjani commented on pkg/endpoint/owner.go:
```diff
@@ -89,6 +89,15 @@ type Owner interface {
+	// label is set on both `add` and `del`, that specific label will exist in the
+	// endpoint's labels.
+	// On success, returns 0, nil. Otherwise, returns an API error code and error msg.
+	UpdateSecLabels(id string, add, del labels.Labels) (int, error)
```
I'm on the fence about this, mostly because UpdateSecLabels (and the various Daemon methods it calls) doesn't seem relevant to the owner of an endpoint. In particular, it seems like this function should be on the endpoint itself but that's a non-trivial refactor from what I can tell :/

joestringer replied:
I agree, but also this PR was getting big enough as it is that I didn't really want to try to tackle this at this stage.

aanm replied:
@joestringer Right, but I also don't understand why is this in the `owner.go`?

joestringer replied:
@aanm Mostly just because Identity management is all in the daemon currently, and the caller isn't in the daemon package so it needs to deal with the abstract class "Owner" rather than knowing the internal details of the daemon. If we refactoring all of the daemon `DeleteIdentity()`, `updateEndpointIdentity()`, etc. stuff out of daemon, then it wouldn't need to be here. We can do that in a follow up PR.

raybejjani reviewed (approved):
I'm not sure about adding UpdateSecLabels to Owner but I don't see avoiding that without a refactor that shouldn't be part of this PR..

joestringer committed:
test: Add basic cilium-health tests

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
prober: Only set up nodes map once.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
prober: Fix issue where not all addresses are probed

Previously, the HTTP prober would typically only attempt to probe one
address based on a random ordering. This commit ensures that all
addresses are probed each time the HTTP prober is run.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+	routes = append(routes, v4Routes...)
+	v6Routes, err := plugins.IPv6Routes(addressing)
+	if err != nil {
+		return fmt.Errorf("Failed to get IPv6 routes")
```
Will there always be an IPv6 and an IPv4 route stored in `addressing`? I ask since in Cilium we have the `Ipv4Disabled` option in `daemon/config.go`

joestringer replied:
I'll loosen this requirement here.

ianvernon commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		log.WithError(err).Fatal("Error while creating cilium-health endpoint")
+	}
+	// FIXME: Should we inherit default config?
```
Default config from where?

joestringer replied:
`SetDefaultOpts()`, ie the defaults for an endpoint for these options:

https://github.com/cilium/cilium/blob/af19d571da430e4642d50ce025998a0ffce9dcc3/pkg/endpoint/endpoint.go#L145-L155

ianvernon commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+	deadline := time.Now().Add(5 * time.Second)
+	for {
+		if _, err := os.Stat(pidfile); err == nil {
+			log.WithField("pidfile", pidfile).Debug("Cilium-health agent running")
```
Keep log messages consistent: Cilium-health --> cilium-health

ianvernon commented on cilium-health/cmd/root.go:
```diff
@@ -82,6 +82,7 @@ func init() {
 	flags.BoolP("json", "j", false, "Format as JSON")
 	// TODO GH #2083 Hide until all commands support JSON output
 	flags.MarkHidden("json")
+	// TODO: Support simplified formatting (Strip RFC3339 in nested run)
```
the logrus formatter should be configurable per package. Can you create a GH issue for this and reference it here? I prefer having all TODO's ref. GH issues.

joestringer replied:
Done: #2567, will link

ianvernon commented on daemon/endpoint.go:
```diff
@@ -675,7 +676,7 @@ func (d *Daemon) UpdateSecLabels(id string, add, del labels.Labels) middleware.R
 			}).WithError(err).Warn("Unable to release temporary identity")
 		}
-		return NewPutEndpointIDLabelsNotFound()
+		return PutEndpointIDLabelsNotFoundCode, fmt.Errorf("")
```
Why an empty error?

joestringer replied:
Apparently because I omitted it ;) I'll fix this up.

ianvernon commented on pkg/node/node_address.go:
```diff
@@ -305,6 +307,26 @@ func SetIPv6Router(ip net.IP) {
+
+// SetIPv6HealthIP sets the IPv6 address of the local cilium-health endpoint
+func SetIPv6HealthIP(ip *net.IP) {
+	ipv6HealthAddress = ip
```
why not just export these variables? that way these functions aren't needed.

joestringer replied:
Just following the existing style.

ianvernon replied:
From discussion with @aanm , we want this convention as it allows us in the future to place mutexes around these objects in the library itself instead of placing the responsibility on the utilizer of these members.

rlenglet commented on api/v1/openapi.yaml:
```diff
@@ -1250,6 +1250,9 @@ definitions:
         items:
           "$ref": "#/definitions/NodeAddressingElement"
+      health-endpoint-address:
+        description: Addresses used for probing cluster connectivity
```
Addresses -> Address

rlenglet commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,213 @@
+	}
+	go func() {
+		in := bufio.NewScanner(stdout)
+		for in.Scan() {
```
It's usually better to keep the output of one command in a single log statement. Otherwise, the output may interleave and make the logs very hard to read. One way to do is to concatenate the lines with newlines or some special character (`"\\n"`) in the same log message.

joestringer replied:
This is reading stdout from a long-running command, using the default `ScanLines` implementation, so each time the cilium-health endpoint logs a message, it will be forwarded to the Cilium log. I'm not sure how we can achieve what you're saying.

aanm replied:
@joestringer is right about the default [`ScanLines`](https://golang.org/pkg/bufio/#ScanLines) implementation. But reading the documentation I understand @rlenglet concerns. The problem might be with "the last non-empty line of input will be returned even if it has no newline" in which case it will log something without a `\n`. I don't think this is a real issues since logrus will print a log message per method call (`Debugf`, `Infof`, etc) even if it ends with or without `\n`

rlenglet replied:
OK, makes sense. LGTM.

rlenglet commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,213 @@
+	go func() {
+		in := bufio.NewScanner(stdout)
+		for in.Scan() {
+			scopedLog.Debug("%s", in.Text())
```
```
scopedLog.Debugf("%s", in.Text())
```
or:
```
scopedLog.Debug(in.Text())
```

joestringer replied:
This trips me up all the time. Having both `Debug()` and `Debugf()` is just asking for people to mis-use the two functions --- particularly since `Debug()` accepts multiple arguments (!). I'd rather consistently use `Debugf()` and not have to care which variation I'm using. (... I'll fix this up)

rlenglet commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,213 @@
+	}
+	go func() {
+		in := bufio.NewScanner(stderr)
+		for in.Scan() {
```
Same comments as above.

rlenglet commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,213 @@
+	if err == nil {
+		routes = append(routes, v4Routes...)
+	} else {
+		log.Debugf("Couldn't get IPv4 routes for health routing")
```
Debugf -> Debug

rlenglet commented on pkg/logging/logfields/logfields.go:
```diff
@@ -56,6 +56,12 @@ const (
 	IPAddr = "ipAddr"
 
+	// V4HealthIP is an address used to contact the cilium-health endpoint
+	V4HealthIP = "v4healthIP"
```
How about `healthIP.IPv4` and `healthIP.IPv6` to be consistent with e.g. `policyID.L3` above?

rlenglet commented on pkg/node/node.go:
```diff
@@ -47,6 +47,14 @@ type Node struct {
+
+	// IPv4HealthIP if set, this is the IPv4 address of the cilium-health
+	// endpoint located on the node.
+	IPv4HealthIP *net.IP
```
`net.IP` is a slice, so it can already take a `nil` value. So do you really need to use a pointer here?

rlenglet commented on pkg/node/node.go:
```diff
@@ -47,6 +47,14 @@ type Node struct {
 	// dev contains the device name to where the IPv6 traffic should be send
 	dev string
+
+	// IPv4HealthIP if set, this is the IPv4 address of the cilium-health
```
if set -> if not nil

rlenglet commented on pkg/node/node_address.go:
```diff
@@ -37,6 +37,8 @@ var (
 	ipv6RouterAddress   net.IP
 	ipv4AllocRange      *net.IPNet
 	ipv6AllocRange      *net.IPNet
+	ipv4HealthAddress   *net.IP
```
No need for pointers. `net.IP` can take a `nil` value.

rlenglet reviewed (approved):
LGTM. Only some non-blocking nitpicks.

aanm commented on cilium-health/cmd/root.go:
```diff
@@ -82,6 +82,7 @@ func init() {
 	flags.BoolP("json", "j", false, "Format as JSON")
 	// TODO GH #2083 Hide until all commands support JSON output
 	flags.MarkHidden("json")
+	// TODO: GH #2567 Simplify log formatting (Strip RFC3339 in nested run)
```
Can you test this again with this PR https://github.com/cilium/cilium/pull/2566?

joestringer replied:
Looks much better now actually. I'll drop that patch

aanm commented on common/plugins/ipam.go:
```diff
@@ -39,6 +39,20 @@ type Route struct {
 
+// ToIPCommand converts the route into a full "ip route ..." command
+func (r *Route) ToIPCommand(dev string) []string {
+	res := []string{"ip"}
```
Should we return `ip` or only the arguments?

Just asking because the caller might decide to execute `ip` on the full path `/bin/ip`

aanm replied:
Also, can you add a unit test to verify the output?

joestringer replied:
It's easy enough to refactor later to allow callers to specify a full path, so I'll leave it as-is for the moment. I've added unit tests for it.

aanm commented on cilium-health/cmd/root.go:
```diff
@@ -133,6 +135,15 @@ func initConfig() {
+	if path := viper.GetString("pidfile"); path != "" {
+		if err := pidfile.Write(path); err != nil {
+			errorString := fmt.Sprintf("Failed to write pidfile %s: %s", path, err.Error())
+			fmt.Println(errorString)
```
Why not `fmt.Printf`?

joestringer replied:
Probably I was avoiding long lines or something. I'll fix this up.

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+		buf.WriteString(cmd)
+		buf.WriteString(join)
+	}
+	buf.Truncate(buf.Len() - len(join))
```
Why did you use `strings.Join(rt.ToIPCommand(dev), " ")` to join strings but not for the `&&` case? :)

joestringer replied:
No particular reason, I can make this more consistent.

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+
+	// XXX: Make the above IPs known to pkg/node
+
+	// XXX: Release IPs later on?
```
Yes please

joestringer replied:
These IPs are now managed in `pkg/node`.

They're allocated once on daemon startup, and never released, just like the other IPs reserved for daemon use like the loopback IP.

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+		if err == nil {
+			return
+		}
+		if err = netlink.LinkDel(veth); err != nil {
```
we have a tool to clean up everything created by cilium. Is this veth cleaned up as well by that tool?

joestringer replied:
Yes, it is prefixed with `cilium_`, and all devices with that prefix are automatically cleaned up.

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+	}()
+
+	// Invoke spawn_netns.sh to spin up the health endpoint
+	pidfile := fmt.Sprintf("%s/health-endpoint.pid", owner.GetStateDir())
```
use [`filepath.Join`](https://golang.org/pkg/path/filepath/#Join)

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+
+	// Invoke spawn_netns.sh to spin up the health endpoint
+	pidfile := fmt.Sprintf("%s/health-endpoint.pid", owner.GetStateDir())
+	os.RemoveAll(pidfile)
```
Out of curiosity, shouldn't we prevent from starting this cilium-health instance if there's one already running? If yes, why are we deleting the old `pidfile`?

joestringer replied:
We won't be able to get logs from the old one. I actually don't expect this condition to be hit, but arguably we should actually read the file if it exists and terminate the process. I'll look into this.

joestringer replied:
I updated this code to open the file and terminate the process.

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+		ip6.String(), ip4.String(), "cilium-health", healthArgs}
+	prog := filepath.Join(owner.GetBpfDir(), "spawn_netns.sh")
+
+	ctx, cancelHealth := context.WithCancel(context.Background())
```
I'm starting to think that we should have a package to run external programs...

tgraf replied:
There is some code in the daemon to run iptables that can be used and factored out.

joestringer replied:
I opened an issue for this: https://github.com/cilium/cilium/issues/2628

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -0,0 +1,211 @@
+	}
+
+	// Wait until the cilium-health endpoint is running before setting up routes
+	deadline := time.Now().Add(5 * time.Second)
```
Are 5 seconds enough?

joestringer replied:
It's pretty arbitrary, do you have another number in mind?

By the way, if this fails then it will only stall cilium daemon startup by max 5 seconds so in some ways it's nice -- we should never expect to hit this but if we do, then at least it only affects the health endpoint.

aanm commented on test/helpers/cilium.go:
```diff
@@ -231,8 +231,8 @@ func (s *SSHMeta) GetEndpoints() *CmdRes {
 func (s *SSHMeta) GetEndpointsIds() (map[string]string, error) {
-	// cilium endpoint list -o jsonpath='{range [*]}{@.container-name}{"="}{@.id}{"\n"}{end}'
-	filter := `{range [*]}{@.container-name}{"="}{@.id}{"\n"}{end}`
+	// cilium endpoint list -o jsonpath='{range [?(@.labels.orchestration-identity[0]!='reserved:health')]}{@.container-name}{"="}{@.id}{"\n"}{end}'
```
Hmm, shouldn't we have a filter? Don't we eventually need to get all endpoints including the `reserved:health` ones?

joestringer replied:
The filter is just made more specific to drop the cilium-health endpoint below. Most people writing tests don't care about the cilium-health endpoint, so it's easier if `GetEndpoints()` keeps that detail away. Otherwise we end up littering a bunch of extra cilium-health endpoint handling throughout the rest of the tests.

For tests that care about the cilium-health endpoint, they can fetch it separately.

aanm commented on tests/helpers.bash:
```diff
@@ -218,7 +218,7 @@ function k8s_num_ready {
-  kubectl -n ${NAMESPACE} exec ${CILIUM_POD} cilium endpoint list | grep $FILTER | grep -v 'not-ready' | grep -c 'ready' || true
+  kubectl -n ${NAMESPACE} exec ${CILIUM_POD} cilium endpoint list | grep $FILTER | grep -v -e 'not-ready' -e 'reserved' | grep -c 'ready' || true
   restore_flag $save "e"
 }
```
Out of curiosity, if I apply a policy, will the `reserved:health` endpoint go into a regenerating state?

joestringer replied:
The intention is that the API cannot modify policy for reserved endpoints, in which case I would expect not. (The actual restrictions on this aren't in the series yet)

aanm commented on pkg/health/server/prober.go:
```diff
@@ -85,6 +85,16 @@ func getPrimaryIP(node *ciliumModels.NodeElement) string {
 }
 
+func getHealthIP(node *ciliumModels.NodeElement) string {
+	if node.HealthEndpointAddress == nil {
```
`if node == nil || ...`

joestringer replied:
This is only used while iterating through a map of nodes which we populate with non-nil objects.. is there a path in mind that you're worried about here?

aanm replied:
Probably not, just worried about if something sets the map with `nil` and we only find out through a panic

aanm commented on test/k8sT/Health.go:
```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 Authors of Cilium
```
missing sign-off in the commit

joestringer replied:
Fixed

aanm commented on test/k8sT/Health.go:
```diff
@@ -0,0 +1,114 @@
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
```
@eloycoto on this case, when is the test cleaned up?

eloycoto replied:
The `cilium_ds.yaml` is use to be in there for all the kubernetes test. No new deployments, pods are in place. So nothing to cleanup.

aanm commented on pkg/health/server/prober.go:
```diff
@@ -346,7 +346,7 @@ func newProber(s *Server, nodes nodeMap) *prober {
 		proberExited: make(chan bool),
```
Missing sign-off

joestringer replied:
Fixed

aanm reviewed (dismissed):
Will I, as an user, be able to change the labels of that special endpoint?

In k8s, what will happen first: the creation of the special endpoint or the k8s api-server connection?

joestringer committed:
daemon: Restrict modification of reserved endpoints

Reserved endpoints should not be able to be modified via the API, so
make some checks in the put/patch/del cases to prevent such calls.

Signed-off-by: Joe Stringer <joe@covalent.io>

borkmann reviewed (approved):
bpf bits lgtm
